edalize/edatool.py

```diff
@@ -310,15 +321,17 @@ def _run_tool(self, cmd, args=[]):
         logger.debug("args  : " + ' '.join(args))
 
         try:
-            subprocess.check_call([cmd] + args,
+            sp_output = subprocess.check_output([cmd] + args,
```

Comment:
@GCHQDeveloper560 I discovered that the following error is given by this:

```
>           raise RuntimeError(_s.format(cmd, e.output.decode()))
E           RuntimeError: '/home/runner/work/edalize/edalize/.tox/py/bin/python' exited with an error code
E           Traceback (most recent call last):
E             File "/tmp/vunit_lesqgqxj/run.py", line 16, in <module>
E               runner = load_runner_hooks().VUnitRunner()
E             File "/tmp/vunit_lesqgqxj/run.py", line 14, in load_runner_hooks
E               return __import__('edalize.vunit_hooks', fromlist=['vunit_hooks'])
E             File "/home/runner/work/edalize/edalize/.tox/py/lib/python3.6/site-packages/edalize/vunit_hooks.py", line 3, in <module>
E               from vunit.ui import Library
E           ModuleNotFoundError: No module named 'vunit'

.tox/py/lib/python3.6/site-packages/edalize/edatool.py:334: RuntimeError
=========================== short test summary info ============================
FAILED tests/test_vunit.py::test_vunit_hooks - RuntimeError: '/home/runner/wo...
========================= 1 failed, 39 passed in 3.96s =========================
ERROR: InvocationError for command /home/runner/work/edalize/edalize/.tox/py/bin/pytest (exited with code 1)
```

IMO this should be fixed as probably check_call may not identify the issue.
This change was needed to suppress the output of `run_tool` and redirect it to log files.

Reply:
It looks like the `vunit_hdl` package isn't getting installed, though I'm not sure why your change is uncovering the error. From tox-dev/tox#13 it appears that tox doesn't pay any attention to the `test_requires` line in `setup.py`. I'm not sure what the correct solution is, but perhaps adding `vunit_hdl` to the `deps` line in `tox.ini` would be the quickest solution.

Reply:
Thanks that worked, but now I see further error:
```
>           raise RuntimeError(_s.format(cmd, e.output.decode()))
E           RuntimeError: '/home/runner/work/edalize/edalize/.tox/py/bin/python' exited with an error code
E             ERROR - Could not detect known LLVM backend by parsing 'ghdl --version'
E           Traceback (most recent call last):
E             File "/home/runner/work/edalize/edalize/.tox/py/lib/python3.5/site-packages/vunit/ui/__init__.py", line 679, in main
E               all_ok = self._main(post_run)
E             File "/home/runner/work/edalize/edalize/.tox/py/lib/python3.5/site-packages/vunit/ui/__init__.py", line 723, in _main
E               return self._main_compile_only()
E             File "/home/runner/work/edalize/edalize/.tox/py/lib/python3.5/site-packages/vunit/ui/__init__.py", line 861, in _main_compile_only
E               simulator_if = self._create_simulator_if()
E             File "/home/runner/work/edalize/edalize/.tox/py/lib/python3.5/site-packages/vunit/ui/__init__.py", line 745, in _create_simulator_if
E               args=self._args, output_path=self._simulator_output_path
E             File "/home/runner/work/edalize/edalize/.tox/py/lib/python3.5/site-packages/vunit/sim_if/ghdl.py", line 74, in from_args
E               backend=cls.determine_backend(prefix),
E             File "/home/runner/work/edalize/edalize/.tox/py/lib/python3.5/site-packages/vunit/sim_if/ghdl.py", line 138, in determine_backend
E               "No known GHDL back-end could be detected from running 'ghdl --version'"
E           AssertionError: No known GHDL back-end could be detected from running 'ghdl --version'
E           Expected to find one of dict_keys(['GCC back-end code generator', 'mcode code generator', 'llvm code generator'])
E           == Output of 'ghdl --version'============================================================
E           
E           =========================================================================================
```

It appears that calling the ghdl tool causes troubles as ghdl is not installed in the environment, hence the assertion.

Reply:
Even with `vunit_hdl` installed `test_vunit` fails on your branch. The first problem seems to be that the GHDL mock command needs to support a `--version` argument. I added one with the following:
```
--- a/tests/mock_commands/ghdl
+++ b/tests/mock_commands/ghdl
@@ -2,6 +2,21 @@
 import os
 import sys

+version = """
+GHDL 0.36-dev (Ubuntu 0.35+git20181129+dfsg-4ubuntu1) [Dunoon edition]
+ Compiled with GNAT Version: 8.3.0
+  GCC back-end code generator
+  Written by Tristan Gingold.
+
+  Copyright (C) 2003 - 2015 Tristan Gingold.
+  GHDL is free software, covered by the GNU General Public License.  There is NO
+  warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+"""
+
+if sys.argv[1] == '--version':
+  print(version)
+  exit(0)
+
 cmd_file = 'analyze.cmd' if sys.argv[1] == '-i' else 'elab-run.cmd'

 with open(cmd_file, 'a') as f:
```

There's still an assertion failing with this change, but I've not looked any deeper.

Reply:
The assertion I am seeing is the following:

```
    def assert_called_once_with(_mock_self, *args, **kwargs):
        """assert that the mock was called exactly once and that that call was
        with the specified arguments."""
        self = _mock_self
        if not self.call_count == 1:
            msg = ("Expected '%s' to be called once. Called %s times." %
                   (self._mock_name or 'mock', self.call_count))
>           raise AssertionError(msg)
E           AssertionError: Expected 'VUnitRunner' to be called once. Called 0 times.

```

I see there are two mock patches called, once with `backend.build()` and the second with `backend.run()`.

Apparently, the backend.build() does not call the VUnitRunner, hence the assertions.

By removing the first mock (with the `backend.build()`, I get an additional error, which is a failure in the vunit tests:

```
E           Compiling into vunit_lib: ../../home/alessandro/.local/lib/python3.6/site-packages/vunit/vhdl/check/src/check.vhd                                  passed
E           Compiling into libx:      ../../data/edalize/tests/test_vunit/minimal/tb_minimal.vhd                                                               passed
E           Compile passed
E           
E           Starting libx.tb_minimal.testcase_1
E           Output file: vunit_out/test_output/libx.tb_minimal.testcase_1_2b9ae7a3e5456ab25a0a139dc55517d9c5e472b6/output.txt
E           fail (P=0 S=0 F=1 T=2) libx.tb_minimal.testcase_1 (0.0 seconds)
E           
E           Starting libx.tb_minimal.testcase_2
E           Output file: vunit_out/test_output/libx.tb_minimal.testcase_2_58710eec74ad7967b1fef6c6efaa30a99be64718/output.txt
E           fail (P=0 S=0 F=2 T=2) libx.tb_minimal.testcase_2 (0.0 seconds)
E           
E           ==== Summary ======================================
E           fail libx.tb_minimal.testcase_1 (0.0 seconds)
E           fail libx.tb_minimal.testcase_2 (0.0 seconds)
E           ===================================================
E           pass 0 of 2
E           fail 2 of 2
E           ===================================================
E           Total time was 0.1 seconds
E           Elapsed time was 0.1 seconds
E           ===================================================
E           Some failed!
```